python-package/lightgbm/basic.py

```diff
@@ -1004,7 +1004,8 @@ def set_reference(self, reference):
         self.set_categorical_feature(reference.categorical_feature)
         self.set_feature_name(reference.feature_name)
         self._set_predictor(reference._predictor)
-        if self.reference is reference:
+        # we're done if self and reference share a common upstrem reference
+        if self.get_ref_chain().intersection(reference.get_ref_chain()):
             return
```

Comment:
my impression is that in this situation, which self.reference and reference share a common ancestor, we would could just not bother with setting `self-reference=reference'.  otherwise we get the same issue that motivated this pr.  perhaps I'm misunderstanding you?